Feature/refactor snakemake impl

pdiercks opened the pull request:
Hi @dglaeser , 
I updated the snakemake workflow as well. I could not find an option to use non-file output (see the notes in the `Snakefile`).
Therefore the rule `parse_stdout` seems pointless, as it would be more natural to pass `config[domain_size]` directly, but just for the sake of illustration, if the value one would like to pass on would be altered by the gmsh job, then this would be the way to implement it I think (I may find a better way later on, but at the moment I don't think non-fie output is possible).

dglaeser commented on simple_use_case/snakemake/Snakefile:
```diff
@@ -1,65 +1,117 @@
+    conda:
+        "../source/envs/preprocessing.yaml"
+    shell:
+        "gmsh -2 -setnumber domain_size {config[domain_size]} {input} -o {output.msh} > {output.stdout}"
```
I actually didn't expose domain_size as an option in the cwl implementation. I thought we had said that the geo file is the input, which spits out the domain size. So I may have to adjust that then...

Or should we simply say geofile & size come as arguments to the workflow, and then we call gmsh with domainsize to create a domain of the desired size, and we simply pass that input value down to the macro-preparation-job instead of reading it from gmsh?

I would say reading it from gmsh's output only makes sense if don't explicitly pass it in. The requirements the wf poses on suitable geo files is different in the two cases. We either expect the geo file be parameterized by a size, or we expect it to tell us the size of the domain that was created... I think both are valid scenario, I just wouldn't mix them.. Which one would you prefer?

pdiercks replied:
okay, I would then prefer to pass domain size to the gmsh job and simply pass it down instead of parsing the output. I thought it be nice to have this as a configuration parameter that you could pass on the command line when executing the workflow. 
The case of the non-file output is then still covered by the number of Dofs.

pdiercks replied:
By the way, should we define `degree` as a configuration parameter as well? It would make sense to me, but is also not that important ...

pdiercks replied:
> So I change cwl again to not read the size from the gmsh output but to receive it as input, actually passing it down to gmsh..

Thanks! 
I will also change this here accordingly. 
> I don't mind. I think the importance is to cover the different types of in/outputs and intermediate data transfer and I think cover that. We can add it or leave it, no strong opinion here. I'd leave that up to you.

I will not add degree as configuration parameter.

pdiercks commented:
Hi @dglaeser , I removed `degree` from the configfile and removed the extra job to parse stdout of the gmsh job. I would like to merge, unless there are still changes required.

dglaeser commented on simple_use_case/snakemake/Snakefile:
```diff
@@ -1,65 +1,86 @@
+        "../source/unit_square.geo"
+    output:
+        msh="preprocessing/square.msh",
+        stdout="preprocessing/gmsh_stdout.txt"
```
is stdout of the mesh step still required?

pdiercks replied:
my bad, I just forgot to remove it. Thanks!

dglaeser commented on simple_use_case/source/poisson.py:
```diff
@@ -88,7 +88,8 @@ def solve_and_write_output(mesh: str, degree: int, outputfile: str, numdofs: str
         "--num-dofs",
-        required=True,
+        required=False,
+        default="dummy_num_dofs.txt",
```
I wonder why this appears as diff, I thought I had already added this to #32 ?

pdiercks replied:
I also wondered about that. This branch was created before the change was made in #32 .
I then rebased this branch to #32.

pdiercks replied:
somehow the commit was duplicated? this is the original one https://github.com/BAMresearch/NFDI4IngScientificWorkflowRequirements/pull/32/commits/c9c76c45f37009f90c43db237666153bed74fbe9 and this is the one in this PR https://github.com/BAMresearch/NFDI4IngScientificWorkflowRequirements/pull/37/commits/d8ff82ad37e9eccf34b1458e02a7751c6d0ce238

pdiercks replied:
I rebased again and force pushed the changes. Do you agree to merge?

dglaeser replied:
looks good, I'll merge
